### DISCUSSION TITLE
Neighbor Location

### DISCUSSION THREAD

**Discussion by eecue** [RhetTbull/osxphotos#905] [Q&A] [answered]
I have a ton of photos that I shot on a DSLR that have no geodata.  However I was also shooting photos with my iPhone at the same time.  Is there an easy way to extract that data and apply it to the nearest neighbor photos?

**Comment by RhetTbull** [comment, marked as answer]
Yes, this should be possible with a relatively simple osxphotos script (though it's not an out of the box command).  I did [something similar](https://github.com/RhetTbull/osxphotos/discussions/495) to add missing location data from Google Takeout (as all Google apps record your location and send it to Google).

When you say "nearest neighbor" I assume you mean chronologically indicating that the DSLR had the time (and timezone?) correctly set?  If so, this should be fairly easy. Here's a rough script that should be close to what you want (you'll need to adjust the "neighbor window").  If you are familiar with python, this should be easy to modify for your needs. If not, I'm happy to help tweak it for you. If you save this as `add_locations.py`, you can run it with osxphotos via `osxphotos run add_locations.py`.

```python
import datetime

import photoscript

import osxphotos

# load photos database
photosdb = osxphotos.PhotosDB()

# go through all photos looking for photos that don't have location
# location is a tuple of (latitude, longitude) or (None, None) if no location
missing_location = [photo for photo in photosdb.photos() if None in photo.location]

# for each photo that's missing location, find nearest neighbor with location
# given some window of time, in this case +/- 1 hour
NEIGHBOR_WINDOW = datetime.timedelta(seconds=60 * 60)  # 1 hour
for photo in missing_location:
    neighbor_start = photo.date - NEIGHBOR_WINDOW
    neighbor_end = photo.date + NEIGHBOR_WINDOW
    # find all photos taken within NEIGHBOR_WINDOW seconds of this photo
    # this is super inefficient as it scans all the photos every time
    # you could optimize by building a sorted list of all photos by datetime
    neighbors = [
        photo
        for photo in photosdb.photos()
        if neighbor_start <= photo.date <= neighbor_end and None not in photo.location
    ]

    if not neighbors:
        print(f"No neighbors found for photo {photo.original_filename} ({photo.date})")
        continue

    nearest = min(neighbors, key=lambda x: abs(x.date - photo.date))
    print(
        f"Nearest neighbor for {photo.original_filename} ({photo.date}) is {nearest.original_filename} ({nearest.date})"
    )
    # set location of photo to location of nearest neighbor
    # osxphotos cannot set location, so use photoscript which can, via AppleScript
    photoscript.Photo(photo.uuid).location = nearest.location

```

**Reply by eecue** [reply]
This is awesome thanks!

**Reply by RhetTbull** [reply]
@eecue In [v0.56.3](https://github.com/RhetTbull/osxphotos/releases/tag/v0.56.3) I've added a new `osxphotos add-locations` command that builds on this idea.  I improved the algorithm so it requires only a single pass through the photos (sorts them all by date then steps forward/backwards as needed to find nearest neighbor). You can specify the window from the command line, e.g.:

  `osxphotos add-locations --window "2 hours" --verbose --dry-run`

Thanks for the inspiration!

**Reply by eecue** [reply]
Awesome!
